File: openrest4cocoa/openrest4cocoa/openrest4cocoa-api/OrderItem.m
//
//  OrderItem.m
//  openrest4cocoa
//
//  Created by Yoav Amit on 5/4/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "OrderItem.h"
#import "Utils.h"

@implementation OrderItem

@synthesize itemId;
@synthesize variations;
@synthesize variationsChoices;
@synthesize comment;
@synthesize price;
@synthesize count;

-(id)init
{
    if ((self = [super init]))
    {
        [self setVariations:[NSArray array]];
        [self setVariationsChoices:[NSArray array]];
        [self setPrice:0];
        [self setCount:1];
    }
    return self;
}

-(id)initWithDictionary:(NSDictionary*)data
{
    if ((self = [self init]))
    {
        [self setItemId:[data valueForKey:@"itemId"]];
  
        if ([data valueForKey:@"variations"] != nil)
        {
            [self setVariations:[Utils refactorJsonArray:[data valueForKey:@"variations"] 
                                               toClass:@"Variation"]];
        }

        if ([data valueForKey:@"variationsChoices"] != nil)
        {
            [self setVariationsChoices:
             [Utils refactorJsonArrayOfArray:[data valueForKey:@"variationsChoices"] 
                                     toClass:@"OrderItem"]];
        }
        
        [self setComment:[data valueForKey:@"comment"]];

        if ([data valueForKey:@"price"] != nil)
        {
            [self setPrice:[[data valueForKey:@"price"] intValue]];
        }
        if ([data valueForKey:@"count"] != nil)
        {
            [self setCount:[[data valueForKey:@"count"] intValue]];
        }
    }
    return self;
}

-(NSDictionary*)proxyForJson
{
    NSMutableDictionary* ret = [NSMutableDictionary dictionaryWithCapacity:0];

    if (itemId != nil) {[ret setValue:itemId forKey:@"itemId"];}
    if (variations != nil) {[ret setValue:variations forKey:@"variations"];}
    if (variationsChoices != nil) {[ret setValue:variationsChoices forKey:@"variationsChoices"];}
    if (comment != nil) {[ret setValue:comment forKey:@"comment"];}
    if (price != 0) {[ret setValue:[NSNumber numberWithInt:price] forKey:@"price"];}
    if (count != 1) {[ret setValue:[NSNumber numberWithInt:count] forKey:@"count"];}
    
    return ret;
}
    
-(int)getTotalPrice
{
    int total = price;
    
    for (int i = 0 ; i < [variationsChoices count] ; i++)
    {
        for (int j = 0 ; j < [[variationsChoices objectAtIndex:i] count] ; j++)
        {
            OrderItem* variation = [[variationsChoices objectAtIndex:i] objectAtIndex:j];
            total += [variation getTotalPrice];
        }
    }
    
    return total*count;
}

-(void)dealloc
{
    [itemId release];
    [variations release];
    [variationsChoices release];
    [comment release];
    [super dealloc];
}

@end

File: openrest4js/openrest4js-api/OrderItem.js
function OrderItemFromObj(data)
{
    var choices = [];
    if (data.variationsChoices)
    {
        for (var i = 0, l = data.variationsChoices.length ; i < l ; i++)
        {
            choices.push(ORJSArrayFromJsonArray(data.variationsChoices[i], OrderItemFromObj));
        }
    }
    return new OrderItem(data.itemId, data.price, data.variations, choices, 
            data.comment, typeof(data.count) === "undefined" ? 1 : data.count);
}

function OrderItem(itemId, price, variations, variationsChoices, comment, count)
{
    this.itemId = itemId;
    this.price = price;
    this.variations = JSON.parse(JSON.stringify(variations || []));
    this.variationsChoices = JSON.parse(JSON.stringify(variationsChoices || []));
    this.comment = comment;
    this.count = count;
}

OrderItem.prototype.toOpenRestObj = function()
{
    var ret = {};

    if (this.itemId) ret['itemId'] = this.itemId;
    if (this.price) ret['price'] = this.price;
    if (this.variations) ret['variations'] = this.variations;
    if (this.variationsChoices) ret['variationsChoices'] = this.variationsChoices;
    if (this.comment) ret['comment'] = this.comment;
    if (this.count) ret['count'] = this.count;

    return ret;
}
